### PR TITLE
Personal/test handler fuspec

### PR DESCRIPTION
**Pull request by Ryzhukhin**
Часть доработок по fuspec perser, 0.1 версия

### CODE REVIEW

File: src/NFun.Fuspec.TestHandler/FuspecTestBase.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            var result = new List<BaseFuspecTestCaseResult>();
+            result.AddRange(_fuspecTestCases.Select(FuspecTestCasesRunner.RunFuspecTest));
+            return result;
+        }
+
+    }
+
+    public static class FuspecTestCasesRunner
+    {
+        public static BaseFuspecTestCaseResult RunFuspecTest(FuspecTestCase testCase)
```
Comment (Ryzhukhin) [inline src/NFun.Fuspec.TestHandler/FuspecTestBase.cs]: Тут идея в том, что проходит по каждому тесту и получает значение, назад отдается только IEnumerable<BaseFuspecTestCaseResult>, где могут быть и плохие тесты, но они помечены как ExceptionTestCaseResult

File: src/NFun.Fuspec.TestHandler/FuspecTestBase.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            var stopwatch = Stopwatch.StartNew();
+            BaseFuspecTestCaseResult testResult;
+            FunRuntime result;
+
+            try
+            {
+                result = FunBuilder.With(testCase.Script).Build();
+                testResult = GoodTestResult(stopwatch);
```
Comment (tmteam) [inline src/NFun.Fuspec.TestHandler/FuspecTestBase.cs]: stopwatch.stop()

File: src/NFun.Fuspec.TestHandler/FuspecTestBase.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            }
+            catch (FunParseException parseException)
+            {
+                testResult = BadTestResult();
+            }
+            catch (FunRuntimeException runtimeException)
+            {
+                testResult = BadTestResult();
```
Comment (tmteam) [inline src/NFun.Fuspec.TestHandler/FuspecTestBase.cs]: How do you want to pass fun-runtime-exception details to bad result? 

Fun runtime exception mean user script bug (that is ok) 
While other exceptions mean nfun bug (that need to be repaired)

It seems you need two bad result types

Comment (Ryzhukhin) [reply src/NFun.Fuspec.TestHandler/FuspecTestBase.cs]: It turns out that it is necessary to show only two types of errors - FunRuntimeException and Exception, already there is ExceptionTestCaseResult, which incorporates all other errors, is this enough or should I add another one?

Comment (tmteam) [reply src/NFun.Fuspec.TestHandler/FuspecTestBase.cs]: At the step of build - you can catch two type of exceptions:
- FunParseException, witch contains errorCode:Enum and additionalErrorInformation:Text
- Other exceptions -  indicator that something totally wrong and there is bug  inside nfun parser

File: src/NFun.Fuspec.TestHandler/FuspecTestCasesRunner.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+            try
+            {
+                //todo parse in and out funResult
+                FunRuntime funResult = FunBuilder.With(testCase.Script).Build();
+                testResult = TestResultConverter.GoodTestResult(stopwatch.ElapsedMilliseconds, testCase);
+            }
+            catch (FunParseException parseException)
+            {
+                testResult = TestResultConverter.BadTestResult(stopwatch.ElapsedMilliseconds, testCase, parseException);
```
Comment (tmteam) [inline src/NFun.Fuspec.TestHandler/FuspecTestCasesRunner.cs]: You need two types of "badTestResult":
- ParseError (when you catch funParseException) with whole fun parse exception information  
- NfunFailure (when you catch all other exception at stage of Build() ) with concrete exception exemplar for further analysis

File: src/NFun.Fuspec.TestHandler/FuspecTestCasesRunner.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            }
+            catch (FunRuntimeException runtimeException)
+            {
+                testResult = TestResultConverter.BadTestResult(stopwatch.ElapsedMilliseconds, testCase, runtimeException);
```
Comment (tmteam) [inline src/NFun.Fuspec.TestHandler/FuspecTestCasesRunner.cs]: You cannot catch FunRuntimeException here. It occurs only at stage of Calculation.
At this step- you need no to handle it

File: src/NFun.Fuspec.TestHandler/Models/BaseFuspecTestCaseResult.cs [hunk 6]
```diff
@@ -5,13 +5,13 @@ namespace NFun.Fuspec.TestHandler.Models
 {
     public class BaseFuspecTestCaseResult
     {
-        public BaseFuspecTestCaseResult(DateTime? totalRunning, FuspecTestCase fuspecTestCase)
+        public BaseFuspecTestCaseResult(long totalRunning, FuspecTestCase fuspecTestCase)
```
Comment (tmteam) [inline src/NFun.Fuspec.TestHandler/Models/BaseFuspecTestCaseResult.cs]: use TimeSpan for totalRunnig

File: src/NFun.Fuspec.TestHandler/Models/BaseFuspecTestCaseResult.cs [hunk 7]
```diff
@@ -5,13 +5,13 @@ namespace NFun.Fuspec.TestHandler.Models
...
         {
             TotalRunning = totalRunning;
             FuspecTestCase = fuspecTestCase;
         }
 
-        public DateTime? TotalRunning { get; set; }
+        public long TotalRunning { get; set; }
```
Comment (tmteam) [inline src/NFun.Fuspec.TestHandler/Models/BaseFuspecTestCaseResult.cs]: Make class immutable. Remove setters from properties

File: src/NFun.Fuspec.TestHandler/TestResultConverter.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using Nfun.Fuspec.Parser.Model;
+using NFun.Fuspec.TestHandler.Models;
+
+namespace NFun.Fuspec.TestHandler
+{
+    public static class TestResultConverter
```
Comment (tmteam) [inline src/NFun.Fuspec.TestHandler/TestResultConverter.cs]: Naming: It is not a converter. Rename it to TestCaseResult

File: src/NFun.Fuspectests/FuspecTestCasesCollection.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+﻿namespace FuspecTests
+{
+    public static class FuspecTestCasesCollection
+    {
+        public static readonly string TestCollection =
```
Comment (tmteam) [inline src/NFun.Fuspectests/FuspecTestCasesCollection.cs]: Use multiline @"..." strings here

File: src/NFun.Fuspectests/FuspecTestParserTests.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        public void FuspecParser_ParseAndRunFuspec_ReturnParsedTests()
+        {
+            //todo пример использования fuspecParser
+            FuspecTestCase[] testCases = GenerateFuspecTestCases(FuspecTestCasesCollection.TestCollection);
+            IEnumerable<BaseFuspecTestCaseResult> testsResult = FuspecTestBase.RunTests(testCases);
+
+            Assert.Pass();
+        }
+
+
```
Comment (tmteam) [inline src/NFun.Fuspectests/FuspecTestParserTests.cs]: Remove spaces

File: src/NFun.Fuspectests/FuspecTestParserTests.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            var stream = new MemoryStream();
+            var writer = new StreamWriter(stream);
+            writer.Write(s);
+            writer.Flush();
+            stream.Position = 0;
+            return stream;
+        }
+
+        private FuspecTestCase[] GenerateFuspecTestCases(string str)
```
Comment (tmteam) [inline src/NFun.Fuspectests/FuspecTestParserTests.cs]: Naming GenerateFuspecTestCases -> ReadTestCases

File: src/NFun.Fuspectests/FuspecTestParserTests.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            GenerateStreamFromString(str);
+            //var specs = new TestCasesReader(new StreamReader(GenerateStreamFromString(str)));
```
Comment (tmteam) [inline src/NFun.Fuspectests/FuspecTestParserTests.cs]: There is something wrong. It smells

File: src/NFun.Fuspectests/FuspecTestParserTests.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+
+
+
+
+
+        
+
+        private static Stream GenerateStreamFromString(string s)
```
Comment (tmteam) [inline src/NFun.Fuspectests/FuspecTestParserTests.cs]: This stack overflow answer is not the best ;)

the best is https://stackoverflow.com/a/38399723/1475183

To use string extensions is much readeable and intuitive

File: src/NFun.Fuspectests/FuspecTestParserTests.cs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+            writer.Flush();
+            stream.Position = 0;
+            return stream;
+        }
+
+        private FuspecTestCase[] GenerateFuspecTestCases(string str)
+        {
+            GenerateStreamFromString(str);
+            //var specs = new TestCasesReader(new StreamReader(GenerateStreamFromString(str)));
+            return new TestCasesReader(new StreamReader(GenerateStreamFromString(str))).Read().TestCases;
```
Comment (tmteam) [inline src/NFun.Fuspectests/FuspecTestParserTests.cs]: Split this bloody line

File: src/NFun.Fuspectests/FuspecTestParserTests.cs [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+namespace NFun.Fuspectests
+{
+    public class FuspecTestParserTests
+    {
+        [Test]
+        public void FuspecParser_ParseAndRunFuspec_ReturnParsedTests()
+        {
+            //todo пример использования fuspecParser
+            FuspecTestCase[] testCases = GenerateFuspecTestCases(FuspecTestCasesCollection.TestCollection);
+            IEnumerable<BaseFuspecTestCaseResult> testsResult = FuspecTestBase.RunTests(testCases);
```
Comment (tmteam) [inline src/NFun.Fuspectests/FuspecTestParserTests.cs]: Use Assert.DoesNotThrow() for such cases

File: src/NFun.Fuspec.TestHandler/TestResultConverter.cs [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+    {
+        public static ExceptionTestCaseResult BadTestResult(long ms, FuspecTestCase testCase, Exception e)
```
Comment (tmteam) [inline src/NFun.Fuspec.TestHandler/TestResultConverter.cs]: Do not use postfix in such situation. just place factories into parent class to use it like this:

TestCaseResult.Bad(testCase, buildTime, e)
TestCaseResult.ParseFailure(testCase, buildTime, e)
TestCaseResut.Passed(testCase, buildTime)

File: src/NFun.Fuspec.TestHandler/FuspecTestCasesRunner.cs [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+{
+    public static class FuspecTestCasesRunner
+    {
+        public static BaseFuspecTestCaseResult RunFuspecTest(FuspecTestCase testCase)
+        {
+            var stopwatch = Stopwatch.StartNew();
+            return ParseTestResult(testCase, stopwatch);
+        }
+
+        private static BaseFuspecTestCaseResult ParseTestResult(FuspecTestCase testCase, Stopwatch stopwatch)
```
Comment (tmteam) [inline src/NFun.Fuspec.TestHandler/FuspecTestCasesRunner.cs]: Why do you pass stopwatch through the argument? Why don't you create and finish it in the same place?

var sw = Stopwatch.StartNew()
DoSomeShit()
sw.Stop()

File: src/NFun.Fuspectests/FuspecTestParserTests.cs (hunk 13 above)
Comment (Ryzhukhin) [reply src/NFun.Fuspectests/FuspecTestParserTests.cs]: @tmteam it's not mine:)) I pulled Natasha's code out of the tests and separated it with special spaces to fix it later

File: src/NFun.Fuspec.TestHandler/FuspecTestCasesRunner.cs (hunk 4 above)
Comment (Ryzhukhin) [reply src/NFun.Fuspec.TestHandler/FuspecTestCasesRunner.cs]: Why do I need two types of errors? The given class has a property with an error, I save the main error, and then it is possible to determine by type what the error was

Comment (tmteam) [reply src/NFun.Fuspec.TestHandler/FuspecTestCasesRunner.cs]: So you make weak typing here to return strict typing in the calling code ;) Why?

You got 3 cases now. It is great if you show it by types. Like enums

File: src/NFun.Fuspec.ConsoleTestHandler/Program.cs [hunk 18]
```diff
@@ -1,12 +1,60 @@
...
         static void Main(string[] args)
         {
+            var ff = new TryFindFuspecTests();
+            ff.FuspecTests();
+            Console.ReadLine();
             Console.WriteLine("Hello World!");
         }
     }
+
+    public class TryFindFuspecTests
```
Comment (tmteam) [inline src/NFun.Fuspec.ConsoleTestHandler/Program.cs]: [Naming] Class is object. It should be noun. For example: FuspecDirectoryReader

File: src/NFun.Fuspec.ConsoleTestHandler/Program.cs [hunk 19]
```diff
@@ -1,12 +1,60 @@
...
+    {
+        private const string FuspecTestExtension = "*.fuspec";
+
+        public void FuspecTests()
```
Comment (tmteam) [inline src/NFun.Fuspec.ConsoleTestHandler/Program.cs]: [Naming] Method should be verb. For ex "ReadFromDirectory"

Comment (tmteam) [reply src/NFun.Fuspec.ConsoleTestHandler/Program.cs]: Class contains no state. It has only one method. Actually it is procedure - so make the class and the method - static

File: src/NFun.Fuspec.ConsoleTestHandler/Program.cs [hunk 20]
```diff
@@ -1,12 +1,60 @@
...
+        {
+            var currentDirectory = Directory.GetCurrentDirectory();
```
Comment (tmteam) [inline src/NFun.Fuspec.ConsoleTestHandler/Program.cs]: [OCP, DI] To use current directory as only one option is not very good. Pass target directory path as a method argument "directoryPath" to allow use reader with any specified path

File: src/NFun.Fuspec.ConsoleTestHandler/Program.cs [hunk 21]
```diff
@@ -1,12 +1,60 @@
...
+            var fuspec = Directory.EnumerateFiles(currentDirectory, FuspecTestExtension, SearchOption.AllDirectories);
+
+            var fuspecFromFiles = fuspec.Select(File.ReadAllText).ToList();
+            var str = string.Join(String.Empty, fuspecFromFiles);
```
Comment (tmteam) [inline src/NFun.Fuspec.ConsoleTestHandler/Program.cs]: Join strings here is useless and causes excessive memory allocation. Imagine you have 100 files, 1Mb each.

it is better to parse each test to FuspecTestCase[] and then join these arrays

File: src/NFun.Fuspec.ConsoleTestHandler/Program.cs [hunk 22]
```diff
@@ -1,12 +1,60 @@
...
+        {
+            var stream = new MemoryStream();
+            var writer = new StreamWriter(stream);
+            writer.Write(s);
+            writer.Flush();
+            stream.Position = 0;
+            return stream;
+        }
+
+        private FuspecTestCase[] GenerateFuspecTestCases(string str)
```
Comment (tmteam) [inline src/NFun.Fuspec.ConsoleTestHandler/Program.cs]: [Naming] Not Generate but Parse

File: src/NFun.Fuspec.ConsoleTestHandler/Program.cs [hunk 23]
```diff
@@ -1,12 +1,60 @@
...
+        {
+            GenerateStreamFromString(str);
+            //var specs = new TestCasesReader(new StreamReader(GenerateStreamFromString(str)));
```
Comment (tmteam) [inline src/NFun.Fuspec.ConsoleTestHandler/Program.cs]: [Commented code]

File: src/NFun.Fuspec.ConsoleTestHandler/Program.cs [hunk 24]
```diff
@@ -1,12 +1,60 @@
...
+            return new TestCasesReader(new StreamReader(GenerateStreamFromString(str))).Read().TestCases;
```
Comment (tmteam) [inline src/NFun.Fuspec.ConsoleTestHandler/Program.cs]: Unreadeable. Create named variable

File: src/NFun.Fuspec.ConsoleTestHandler/Program.cs [hunk 25]
```diff
@@ -1,12 +1,60 @@
...
+
+        public void FuspecTests()
+        {
+            var currentDirectory = Directory.GetCurrentDirectory();
+            var fuspec = Directory.EnumerateFiles(currentDirectory, FuspecTestExtension, SearchOption.AllDirectories);
+
+            var fuspecFromFiles = fuspec.Select(File.ReadAllText).ToList();
+            var str = string.Join(String.Empty, fuspecFromFiles);
+            FuspecTestCase[] testCases = GenerateFuspecTestCases(str);
+            IEnumerable<BaseFuspecTestCaseResult> testsResult = FuspecTestBase.RunTests(testCases);
```
Comment (tmteam) [inline src/NFun.Fuspec.ConsoleTestHandler/Program.cs]: You need to print results on a console. You may return this testResults and add method "Print results"

**Pull request closed without merging**